Re-create the OVN DB with the k8s pod annotation, should it exists.

alexanderConstantinescu opened the pull request:
This had been done previously but was removed with commit https://github.com/ovn-org/ovn-kubernetes/commit/f99a10b1a4b9fa6fefb27e60628cc0960921ca5e 

We however still need this specific behaviour in case the OVN DB is completely destroyed, but the cluster and the corresponding kubernetes objects remain intact (possibly during a roll-out from OVN -> OVN HA). If not we'll have a different state between k8s vs. OVN. 

Test cases updated to test for this behaviour.

girishmg commented:
@dcbw  @alexanderConstantinescu the new code in pods.go will never be exercised unless I have misunderstood something. With the `--ha` option it was getting exercised, however we torched that option and therefore the code.

dcbw commented:
@girishmg if the OVN database is blown away and ovnkube master restarted, pods will still exist on the nodes and in the Kube API with the previous IP and MAC address. But when the master is restarted and the WatchPods handler is called for each existing pod (which happens when the handler is connected as expected for a shared informer) it would previously re-allocate an IP address and set the annotation. But because the pod on each node would not be restarted, the OVN annotation and the Kube API would have different IP addresses.

This PR fixes that issue and I think it's necessary even if the database is not empty on master restart. If for any reason the master gets restarted, when it syncs pods it *must* use the IP/MAC from the Kube API as that is our single source of truth. And that means parsing the previous pod annotation and using that as a static IP address, instead of letting OVN reallocate with a dynamic address.

dcbw commented on go-controller/pkg/ovn/common.go:
```diff
@@ -199,16 +199,20 @@ func (oc *Controller) deletePortGroup(hashName string) {
+	return ovnAnnotationMap["mac_address"]
+}
+
+func (oc *Controller) getIPFromOvnAnnotation(ovnAnnotation string) string {
```
@alexanderConstantinescu let's return an error from this function if the IP doesn't pass `net.ParseCIDR()`

dcbw commented on go-controller/pkg/ovn/common.go:
```diff
@@ -199,16 +199,20 @@ func (oc *Controller) deletePortGroup(hashName string) {
 
-func (oc *Controller) getIPFromOvnAnnotation(ovnAnnotation string) string {
-	if ovnAnnotation == "" {
+func (oc *Controller) getMacFromOvnAnnotation(ovnAnnotation string) string {
```
@alexanderConstantinescu let's return an error from this function if the MAC doesn't pass `net.ParseMAC()`

dcbw commented on go-controller/pkg/ovn/common.go:
```diff
@@ -221,6 +225,21 @@ func (oc *Controller) getIPFromOvnAnnotation(ovnAnnotation string) string {
 	return ipAddressMask[0]
 }
 
+func (oc *Controller) getAnnotationMap(ovnAnnotation string) (map[string]string, error) {
```
@alexanderConstantinescu instead of this function, we can now use:

`annotation, err := util.UnmarshalPodAnnotation(pod.Annotations["ovn"])`

alexanderConstantinescu replied:
Thanks for the comment. I have modified the code accordingly. Just wanted to raise the attention to the fact that: `UnmarshalPodAnnotation` returns the IP address in CIDR format, where as `getIPFromOvnAnnotation` did not. I am not sure which is the right format to use when passing this off to the OVN ctl binaries, or if that even matters to them?

dcbw commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -238,19 +238,44 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) {
-			portName, out, stderr, err)
-		return
+	oldAnnotation, isStaticIP := pod.Annotations["ovn"]
+
```
@alexanderConstantinescu I would do something like this to validate that the existing annotation is OK:

```
annotation, err := util.UnmarshalPodAnnotation(pod.Annotations["ovn"])
if err == nil {
		out, stderr, err = util.RunOVNNbctl("--may-exist", "lsp-add",
			logicalSwitch, portName, "--", "lsp-set-addresses", portName,
			annotation.MAC + " " + annotation.IP, "--", "set",
			"logical_switch_port", portName,
			"external-ids:namespace="+pod.Namespace,
			"external-ids:logical_switch="+logicalSwitch,
			"external-ids:pod=true", "--", "--if-exists",
			"clear", "logical_switch_port", portName, "dynamic_addresses")
		if err != nil {
			logrus.Errorf("Failed to add logical port to switch "+
				"stdout: %q, stderr: %q (%v)",
				out, stderr, err)
			return
		}
} else {
     <do existing stuff>
}
```

girishmg commented:
> @girishmg if the OVN database is blown away and ovnkube master restarted, pods will still exist on the nodes and in the Kube API with the previous IP and MAC address. But when the master is restarted and the WatchPods handler is called for each existing pod (which happens when the handler is connected as expected for a shared informer) it would previously re-allocate an IP address and set the annotation to the new address. But because the pod on each node would not be restarted, the OVN annotation and the Kube API would have different IP addresses.
> 
> This PR fixes that issue and I think it's necessary even if the database is not empty on master restart. If for any reason the master gets restarted, when it syncs pods it _must_ use the IP/MAC from the Kube API as that is our single source of truth. And that means parsing the previous pod annotation and using that as a static IP address, instead of letting OVN reallocate with a dynamic address.

@dcbw I understand. However, if OVN DBs are corrupted then we have much bigger issue. We will need to resurrect various logical resources.

As I understand, the thinking was to build the whole OVN DB from the metadata that we would have stashed in Pod and Node annotation (our source of truth). I am working on adding support for a syncThread that will synchronize the data from etcd datastore to OVN DB. I was waiting on the management port creation to move to master and OVN GW creation to master.

This syncThread should handle the OVN DB being lost or getting corrupted and such.

alexanderConstantinescu commented:
@dcbw I took your comments into account and updated the PR. I however took the liberty and changed all instances of `getIPFromOvnAnnotation` to the new method `util.UnmarshalPodAnnotation`. I thought it would be better to remove all together to avoid confusion between which method to use in the future. I am however unsure if this is right due to the CIDR format difference between these methods. Could you please have a look at my update and advise? I could revert this additional change in that case.

dcbw commented:
@alexanderConstantinescu can you rebase? I merged another PR that does most of the UnmarshalPodAnnotation stuff, so some of that may be able to go away.

alexanderConstantinescu commented:
@dcbw Rebase done. Please have another look

dcbw commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -245,19 +245,41 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) {
+	if err == nil {
+		out, stderr, err = util.RunOVNNbctl("--may-exist", "lsp-add",
+			logicalSwitch, portName, "--", "lsp-set-addresses", portName,
+			fmt.Sprintf("%s %s", annotation.MAC, annotation.IP), "--", "set",
```
@alexanderConstantinescu I think we need to do `annotation.IP.IP` here. Otherwise it'll write "1.2.3.4/24" which isn't what we expect in GetPortAddresses().

alexanderConstantinescu replied:
Done, thanks for the remark on that.

dcbw commented:
@alexanderConstantinescu one comment, otherwise good.
